### PR TITLE
add api requirements for services

### PR DESCRIPTION
**Pull request by GuillaumeDSM**

### CODE REVIEW

File: octobot_trading/api/exchange.py [hunk 1]
```diff
@@ -32,3 +32,15 @@ def create_new_exchange(config, exchange_name,
                            is_collecting=is_collecting,
                            exchange_only=exchange_only,
                            backtesting_files=backtesting_files)
+
+
+def get_trading_pairs(exchange_manager):
```
Comment (Herklos) [inline octobot_trading/api/exchange.py]: How can a module get the exchange_manager to call this API ?

File: octobot_trading/api/orders.py [hunk 2]
```diff
@@ -36,3 +40,23 @@ def get_open_orders(exchange_manager, symbol: str):
...
+
+async def cancel_all_open_orders(exchange_manager):
+    return await exchange_manager.trader.cancel_all_open_orders()
+
+
+async def cancel_all_open_orders_with_currency(exchange_manager, currency):
+    return await exchange_manager.trader.cancel_all_open_orders_with_currency(currency)
+
+
+async def cancel_order_from_description(exchange_manager, order_description):
```
Comment (Herklos) [inline octobot_trading/api/orders.py]: I didn't like the implementation of OctoBot < 0.3.7 and removed it.

Now each order (simulated or real) has an order_id.

So this method should be cancel_order_from_id

File: octobot_trading/api/orders.py [hunk 3]
```diff
@@ -36,3 +40,23 @@ def get_open_orders(exchange_manager, symbol: str):
                                                                                                  current_price=current_price,
                                                                                                  quantity=quantity,
                                                                                                  price=price))
+
+
+def get_open_orders(exchange_manager):
```
Comment (Herklos) [inline octobot_trading/api/orders.py]: :+1:

File: octobot_trading/api/profitability.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+
+
+def get_profitability_stats(exchange_manager):
```
Comment (Herklos) [inline octobot_trading/api/profitability.py]: :+1:

File: octobot_trading/api/trader.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+from octobot_trading.api import LOGGER_TAG
+from octobot_commons.logging.logging_util import get_logger
+
+LOGGER = get_logger(LOGGER_TAG)
+
+
+def is_trader_enabled_in_config(exchange_manager):
```
Comment (Herklos) [inline octobot_trading/api/trader.py]: Can you add some typing ?
`def is_trader_enabled_in_config(exchange_manager) -> bool`
`def is_trader_enabled(exchange_manager) -> bool`

etc...

File: octobot_trading/api/trader.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+def set_trading_enabled(exchange_manager, enabled):
+    exchange_manager.trader.is_enabled = enabled
+
+
+def is_trader_simulated(exchange_manager):
+    return exchange_manager.is_trader_simulated
+
+
+def get_trader_risk(exchange_manager):
```
Comment (Herklos) [inline octobot_trading/api/trader.py]: Can you add some typing ?
`def get_trader_risk(exchange_manager) -> float`
`def set_trader_risk(exchange_manager, risk: float)`
etc...

File: octobot_trading/api/trader.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+
+def set_trader_risk(exchange_manager, risk):
+    return exchange_manager.trader.set_risk(risk)
+
+
+async def sell_all_everything_for_reference_market(exchange_manager):
+    LOGGER.error("from sell_all_everything_for_reference_market: "
+                 "exchange_manager.trader.sell_all_currencies is not implemented yet")
+    # TODO: uncomment when implemented
```
Comment (Herklos) [inline octobot_trading/api/trader.py]: Should be soon implemented ;) #27

File: octobot_trading/api/trades.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    if symbol is None:
+        return exchange_manager.exchange_personal_data.trades_manager.values()
+    else:
+        return [trade
+                for trade in exchange_manager.exchange_personal_data.trades_manager.values()
+                if trade.symbol == symbol]
+
+
+def get_total_paid_trading_fees(exchange_manager):
+    return exchange_manager.exchange_personal_data.trades_manager.get_total_paid_fees()
```
Comment (Herklos) [inline octobot_trading/api/trades.py]: A new line at the end of the file is missing ;)

File: octobot_trading/data_manager/trades_manager.py [hunk 9]
```diff
@@ -49,6 +49,16 @@ def upsert_trade_instance(self, trade):
             self.trades[trade.order_id] = trade
             self._check_trades_size()
 
+    def get_total_paid_fees(self):
```
Comment (Herklos) [inline octobot_trading/data_manager/trades_manager.py]: :+1:

Comment (Herklos) [inline octobot_trading/data_manager/trades_manager.py]: Should be declared in `trades_manager.pxd`

File: octobot_trading/api/exchange.py (hunk 1 above)
Comment (GuillaumeDSM) [reply octobot_trading/api/exchange.py]: From ```create_new_exchange``` and ```get_exchange_manager_from_exchange_name``` that will be pushed with the next push :)

File: octobot_trading/api/orders.py (hunk 2 above)
Comment (GuillaumeDSM) [reply octobot_trading/api/orders.py]: Good !

File: octobot_trading/api/exchange.py (hunk 1 above)
Comment (Herklos) [reply octobot_trading/api/exchange.py]: Nice !

**Comment by Herklos** [comment]
You need to have https://github.com/Drakkar-Software/OctoBot-Commons/pull/8 merged to have the CI to pass right ?

**Comment by GuillaumeDSM** [comment]
> You need to have [Drakkar-Software/OctoBot-Commons#8](https://github.com/Drakkar-Software/OctoBot-Commons/pull/8) merged to have the CI succeeding right ?

Right, we need the new OctoBot-Commons version

File: octobot_trading/api/exchange.py [hunk 10]
```diff
@@ -46,6 +45,10 @@ def get_exchange_names() -> list:
     return Exchanges.instance().get_exchange_names()
 
 
+def get_exchange_channel(chan_name, exchange_name) -> ExchangeChannel:
```
Comment (Herklos) [inline octobot_trading/api/exchange.py]: How this API method is used ?

Comment (GuillaumeDSM) [reply octobot_trading/api/exchange.py]: To subscribe to order channels in : https://github.com/Drakkar-Software/OctoBot-Notifications/pull/2
```python
    async def _subscribe_to_order_channels(self):
        order_channel_name = OrdersChannel.get_name()
        try:
            for exchange_name in get_exchange_names():
                channel = get_exchange_channel(order_channel_name, exchange_name)
                await channel.new_consumer(self._order_notification_callback)
        except KeyError:
            self.logger.error("No order channel to subscribe to: impossible to send order notifications")
```

Comment (GuillaumeDSM) [reply octobot_trading/api/exchange.py]: We could also add an API to subscribe to the order channel with a callback, that would wrap the internal wiring of octobot-trading but would imply a new API for each other future subscription

Comment (Herklos) [reply octobot_trading/api/exchange.py]: I really don't like to interact with exchange channel outside of OctoBot-Trading.

Can we imagine to pass the callback as the parameter of the API method like :
```
def subscribe_to_order_channels(callback):
    for exchange_name in get_exchange_names():
        channel = get_exchange_channel(order_channel_name, exchange_name)
        await channel.new_consumer(self._order_notification_callback) 
```

Comment (GuillaumeDSM) [reply octobot_trading/api/exchange.py]: Yep, I'll add that

File: octobot_trading/modes/abstract_trading_mode.py [hunk 11]
```diff
@@ -40,7 +40,7 @@ def __init__(self, config, exchange_manager):
 
     @classmethod
     def get_name(cls) -> str:
-        return __class__.__name__
+        return cls.__name__
```
Comment (Herklos) [inline octobot_trading/modes/abstract_trading_mode.py]: :+1:

File: octobot_trading/traders/trader.py [hunk 12]
```diff
@@ -149,10 +149,20 @@ def set_risk(self, risk):
...
-                                                                                           cancelled_order)
+                # skip_upsert when cancelled_order is None (nothing to update because simulated order)
+                skip_upsert = cancelled_order is None
+                cancelled_order = odr if skip_upsert else cancelled_order
+                await self.exchange_manager.exchange_personal_data.handle_order_update(order.symbol,
+                                                                                       order.order_id,
+                                                                                       cancelled_order,
+                                                                                       skip_upsert=skip_upsert)
+
+    async def cancel_order_with_id(self, order_id):
```
Comment (Herklos) [inline octobot_trading/traders/trader.py]: :+1:

File: octobot_trading/traders/trader.py [hunk 13]
```diff
@@ -254,7 +264,9 @@ def set_risk(self, risk):
 
     # TODO : should use updater methods
     async def force_refresh_orders_and_portfolio(self, portfolio=None, delete_desync_orders=True):
-        pass
+        # await self.exchange_manager.reset_web_sockets_if_any()    # Might now be useless (auto reconnect), to confirm
```
Comment (Herklos) [inline octobot_trading/traders/trader.py]: Yes, should not be managed here. It's now self-managed by the websocket feed.

File: octobot_trading/traders/trader.py [hunk 14]
```diff
@@ -254,7 +264,9 @@ def set_risk(self, risk):
...
+        await self.force_refresh_orders(portfolio, delete_desync_orders=delete_desync_orders)
+        await self.force_refresh_portfolio(portfolio)
```
Comment (Herklos) [inline octobot_trading/traders/trader.py]: This methods is not declared isn't it ? (commented below)

File: octobot_trading/exchanges/data/exchange_personal_data.py [hunk 15]
```diff
@@ -99,20 +99,23 @@ def __init__(self, exchange_manager):
...
+    async def handle_order_update(self, symbol, order_id, order, should_notify: bool = True,
+                                  skip_upsert: bool = False) -> (bool, bool):
         try:
-            changed: bool = self.orders_manager.upsert_order(order_id, order)
+            changed: (bool, bool) = (False, False)
+            if not skip_upsert:
+                changed = self.orders_manager.upsert_order(order_id, order)
             if should_notify:
                 await get_chan(ORDERS_CHANNEL, self.exchange.name).get_internal_producer() \
-                    .send_(symbol=symbol,
```
Comment (Herklos) [inline octobot_trading/exchanges/data/exchange_personal_data.py]: :+1:

**Pull request merged**